### PR TITLE
Fix a bug in module loading when package.json's main does not resolve.

### PR DESCRIPTION
**Pull request by nicks**
Fixes https://github.com/joyent/node/issues/9118.

Before this change, the test would load the wrong module 'b'.

### CODE REVIEW

**Comment by jasnell** [comment]
@cjihrig @trevnorris @misterdjules ... thoughts on this one?

**Comment by trevnorris** [comment]
No comments from me.

/cc @othiym23 Any comment from the npm side of how this would affect things?

**Comment by othiym23** [comment]
I don't think this will affect npm too much, because it already insists that `main` resolve as part of publishing. That said, this could have all sorts of weird breaking effects in practice, so even though it's arguably a by fix, it should probably be tagged semver-major.

**Comment by jasnell** [comment]
Ok, given that, this would need to land in nodejs/node if it lands at all. Going to close this off here. @nicks .. if you'd like to pursue this further, may I ask that you open a new PR against nodejs/node master?

**Comment by nicks** [comment]
sure, will do. one moment

**Comment by nicks** [comment]
opened a PR as https://github.com/nodejs/node/pull/2568

File: lib/module.js [hunk 1]
```diff
@@ -91,13 +91,17 @@ function statPath(path) {
...
+}
+
 function readPackage(requestPath) {
   if (hasOwnProperty(packageMainCache, requestPath)) {
     return packageMainCache[requestPath];
   }
 
   try {
-    var jsonPath = path.resolve(requestPath, 'package.json');
+    var jsonPath = getPackageJsonPath(requestPath);
```
Comment (mscdex) [inline lib/module.js]: Why is this change necessary?
